plugins/ingest/src/main/java/org/elasticsearch/plugin/ingest/PipelineStore.java

```diff
@@ -74,29 +75,36 @@
     private volatile Map<String, PipelineDefinition> pipelines = new HashMap<>();
 
     @Inject
-    public PipelineStore(Settings settings, Provider<Client> clientProvider, ThreadPool threadPool, Environment environment, ClusterService clusterService, Map<String, Processor.Factory> processors) {
+    public PipelineStore(Settings settings, Provider<Client> clientProvider, ThreadPool threadPool, Environment environment, ClusterService clusterService, Map<String, ProcessorFactoryProvider> processorFactoryProviders) {
         super(settings);
         this.threadPool = threadPool;
         this.clusterService = clusterService;
         this.clientProvider = clientProvider;
         this.scrollTimeout = settings.getAsTime("ingest.pipeline.store.scroll.timeout", TimeValue.timeValueSeconds(30));
         this.pipelineUpdateInterval = settings.getAsTime("ingest.pipeline.store.update.interval", TimeValue.timeValueSeconds(1));
-        for (Processor.Factory factory : processors.values()) {
-            factory.setConfigDirectory(environment.configFile());
+        Map<String, Processor.Factory> processorFactories = new HashMap<>();
+        List<Closeable> closeables = new ArrayList<>();
+        for (Map.Entry<String, ProcessorFactoryProvider> entry : processorFactoryProviders.entrySet()) {
+            Processor.Factory processorFactory = entry.getValue().get(environment);
+            processorFactories.put(entry.getKey(), processorFactory);
+            if (processorFactory instanceof Closeable) {
+                closeables.add((Closeable) processorFactory);
+            }
         }
-        this.processorFactoryRegistry = Collections.unmodifiableMap(processors);
-        clusterService.add(new PipelineStoreListener());
+        // TODO: When org.elasticsearch.node.Node can close Closable instance we should delegate to that or something like that
+        // Ideally we should not do this here, but when a node gets stopped closable factores need to get closed too:
         clusterService.addLifecycleListener(new LifecycleListener() {
```

Comment:
It used to. During code review [1] it was noted that the PipelineStore should be made lighter for testing purposes.

1: https://github.com/s1monw/elasticsearch/commit/042ef8c20476459f92ea1c2ea12bc1950211e693#diff-c69f941ab536509edea2ed0715c9a0adR63